### PR TITLE
jv-registration-validation

### PR DESCRIPTION
**Pull request by OleksiiMakovkin**
created test class RegistrationServiceImplTest

### CODE REVIEW

File: src/main/java/core/basesyntax/CustomException.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package core.basesyntax;
+
+public class CustomException extends RuntimeException {
```
Comment (aaananas) [inline src/main/java/core/basesyntax/CustomException.java:3]: change to a meaningful name

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 2]
```diff
@@ -1,14 +1,49 @@
...
+        if (storageDao.get(user.getLogin()) != null) {
+            throw new CustomException("Such user exists");
+        }
+        if (!Character.isLetter(user.getLogin().charAt(0))) {
+            throw new CustomException("Login must start with a letter");
+        }
+        if (user.getPassword() == null) {
+            throw new CustomException("User password can't not be null");
+        }
+        if (user.getPassword().equals("")) {
```
Comment (aaananas) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:32]:
```suggestion
        if (user.getPassword().isEmpty()) {
```

**Review by aaananas — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 3]
```diff
@@ -1,14 +1,49 @@
...
+        if (user.getPassword().length() > MAX_LENGTH_PASSWORD) {
+            throw new RegistrationException("You entered too long password");
+        }
+        if (user.getAge() < MINIMAL_AGE) {
+            throw new RegistrationException("You are too small)");
+        }
+        if (user.getAge() > MAX_AGE) {
+            throw new RegistrationException("You are too old");
+        }
+        return user;
```
Comment (aaananas) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:47]: you dont save the user in dao

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    /*
+    Testing the login field
+     */
+
+    @Test
+    void register_loginDetailsAreCorrect_Ok() {
+        user.setLogin(DEFAULT_LOGIN);
+        User actual = registrationService.register(user);
+        assertEquals(user, actual);
```
Comment (aaananas) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:47]: you should also check if this user is in the storage

**Review by aaananas — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        User expected = storageDao.get(DEFAULT_LOGIN);
+        assertEquals(expected, actual);
+
+    }
+
+    @Test
+    void register_user_notOk() {
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:54]:
```suggestion
    void register_userAlreadyExists_notOk() {
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        Storage.people.add(user);
+        assertThrows(RegistrationException.class, () ->
+                registrationService.register(user));
+    }
+
+    @Test
+    void register_withNull_notOk() {
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:61]:
```suggestion
    void register_userIsNull_notOk() {
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    @Test
+    void register_emptyLineLogin_notOk() {
+        user.setLogin("  ");
+        assertThrows(RegistrationException.class, () ->
+                registrationService.register(user));
+    }
+
+    @Test
+    void register_checkForAbsenceInTheFirstLetterOfCharacter_notOk() {
+        user.setLogin("#oleksii");
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:75]: checkForAbsenceInTheFirstLetterOfCharacter - I can not understand this
What case are you testing and why there is exception?

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    }
+    /*
+    Testing the password field
+     */
+
+    @Test
+    void register_withCorrectPassword_ok() {
+        user.setPassword(DEFAULT_PASSWORD);
+        User actual = registrationService.register(user);
+        assertEquals(user, actual);
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:87]: same as the first test. Delete it

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    @Test
+    void register_withNullInPassword_notOk() {
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:91]:
```suggestion
    void register_nullInPassword_notOk() {
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        user.setPassword(null);
+        assertThrows(RegistrationException.class, () ->
+                registrationService.register(user));
+    }
+
+    @Test
+    void register_withEmptyFiled_notOk() {
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:98]: withEmptyFiled - which one field?

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        user.setPassword("  ");
+        assertThrows(RegistrationException.class, () ->
+                registrationService.register(user));
+    }
+
+    @Test
+    void register_withVeryLongPassword_notOk() {
+        user.setPassword("1234567891011121314");
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:106]: What is the longest acceptable length according to the task?

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        assertThrows(RegistrationException.class, () ->
+                registrationService.register(user));
+    }
+
+    @Test
+    void register_withVerySmallPassword_notOk() {
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:112]:
```suggestion
    void register_shortPassword_notOk() {
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+        user.setPassword("12");
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:113]: always test edge cases

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+        assertThrows(RegistrationException.class, () ->
+                registrationService.register(user));
+    }
+    /*
+    Testing the age field
+     */
+
+    @Test
+    void register_underageUser_notOk() {
+        user.setAge(12);
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:123]: always test edge cases

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+        user.setPassword("12");
+        assertThrows(RegistrationException.class, () ->
+                registrationService.register(user));
+    }
+    /*
+    Testing the age field
+     */
+
+    @Test
+    void register_underageUser_notOk() {
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:122]:
```suggestion
    void register_lowUserAge_notOk() {
```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+        user.setAge(12);
+        assertThrows(RegistrationException.class, () ->
+                registrationService.register(user));
+    }
+
+    @Test
+    void register_veryOldUser_notOk() {
+        user.setAge(90);
```
Comment (Rommelua) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:130]: What is the max acceptable age according to the task?

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java (hunk 16 above)
Comment (OleksiiMakovkin) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:130]: the task does not specify the maximum age, I myself decided to add this test, if we follow strictly according to the task, then I delete it

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java (hunk 7 above)
Comment (OleksiiMakovkin) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:75]: here I wanted to check that the first element is a letter, and by exception they didn’t come up with anything better in all tests

Comment (andriy-shulepa) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:75]: then name it like register_firstLoginCharacterIsNotLetter_notOk

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+
+    @Test
+    void register_withEmptyFiled_notOk() {
+        user.setPassword("  ");
+        assertThrows(RegistrationException.class, () ->
+                registrationService.register(user));
+    }
+
+    @Test
+    void register_shortPassword_notOk() {
```
Comment (andriy-shulepa) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:98]: Looks like long password. rename test

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java (hunk 12 above)
Comment (andriy-shulepa) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:105]: not fixed. apply suggested changes

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java (hunk 13 above)
Comment (andriy-shulepa) [reply src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:106]: not fixed. test password which is one character smaller than allowed length

**Review by andriy-shulepa — CHANGES_REQUESTED** [review]

**Review by andriy-shulepa — APPROVED** [review]
